File: storage/data_original/at_coder/agc028/A/4593963.java
import java.util.*;
import java.io.*;
import java.math.*;

public class Main {

	void solve (FastScanner in, PrintWriter out) {

		int n = in.nextInt(), m = in.nextInt();
		String s = in.next(), t = in.next();
		
		int gcd = (int)gcd(n, m);
		int i = 0;
		while (true) {
			int a = i*n/gcd, b = i*m/gcd;
			
			if (a>=n || b>=m) break;
			if (s.charAt(a) != t.charAt(b)) {
				out.println(-1);
				return;
			}
			
			i++;
		}
		
		out.println(lcm(n, m));
		

	}

	static long gcd (long a, long b) {return b>0?gcd(b,a%b):a;}
	static long lcm (long a, long b) {return a/gcd(a,b)*b;}
	
	
	void print (Object... ar) {
		System.out.println(Arrays.deepToString(ar));
	}

	public static void main(String[] args) {
		FastScanner in = new FastScanner(System.in);
		PrintWriter out = new PrintWriter(System.out);
		Main main = new Main();
		main.solve(in, out);
		in.close();
		out.close();
	}

	static class FastScanner {

		private InputStream in;
		private byte[] buffer = new byte[1024];
		private  int length = 0, p = 0;

		public FastScanner (InputStream stream) {
			in = stream;
		}
		public boolean hasNextByte () {
			if (p < length) return true;
			else {
				p = 0;
				try {length = in.read(buffer);}
				catch (Exception e) {e.printStackTrace();}
				if (length <= 0) return false;
			}
			return true;
		}
		public int readByte () {
			if (hasNextByte() == true) return buffer[p++];
			return -1;
		}
		public boolean isPrintable (int n) {return 33<=n&&n<=126;}
		public void skip () {
			while (hasNextByte() && !isPrintable(buffer[p])) p++;
		}
		public boolean hasNext () {skip(); return hasNextByte();}
		public String next () {
			if (!hasNext()) throw new NoSuchElementException();
			StringBuilder sb = new StringBuilder();
			int t = readByte();
			while (isPrintable(t)) {
				sb.appendCodePoint(t);
				t = readByte();
			}
			return sb.toString();
		}
		public String[] nextArray (int n) {
			String[] ar = new String[n];
			for (int i=0; i<n; i++) ar[i] = next();
			return ar;
		}
		public int nextInt () {return Math.toIntExact(nextLong());}
		public int[] nextIntArray (int n) {
			int[] ar = new int[n];
			for (int i=0; i<n; i++) ar[i] = nextInt();
			return ar;
		}
		public long nextLong () {
			if (!hasNext()) throw new NoSuchElementException();
			boolean minus = false;
			int temp = readByte();
			if (temp == '-') {
				minus = true;
				temp = readByte();
			}
			if (temp<'0' || '9'<temp) throw new NumberFormatException();
			long n = 0;
			while (isPrintable(temp)) {
				if ('0'<=temp && temp<='9') {
					n *= 10;
					n += temp - '0';
				}
				else throw new NumberFormatException();
				temp = readByte();
			}
			return minus? -n : n;
		}
		public double nextDouble () {
			return Double.parseDouble(next());
		}
		public void close () {
			try {in.close();}
			catch(Exception e){}
		}
	}

}

File: data/data_transpiled/at_coder/agc028/A/4593963.py
def _import ( ) : return _import ( )
import math
import os
class Main ( object ) :
    def solve ( self , stream ) :
        n , m = next ( stream )
        s , t = next ( stream )
        gcd = int ( gcd ( n , m ) )
        i = 0
        while True :
            a , b = i * n // gcd , i * m // gcd
            if a >= n or b >= m :
                raise StopIteration
            if s [ a ] != t [ b ] :
                stream.write ( - 1 )
                return
            i += 1
        stream.write ( lcm ( n , m ) )
    def gcd ( a , b ) :
        return gcd ( b , a % b ) if b > 0 else a
    def lcm ( a , b ) :
        return a // gcd ( a , b ) * b
    def write ( self , * ar ) :
        print ( "".join ( map ( lambda x : x + "" , ar ) ) )
    def main ( self ) :
        stream = sys.stdin
        stream.write ( "\n" )
        self.main = Main ( )
        self.solve ( stream , stream )
        stream.close ( )
        stream.close ( )
    class FastScanner ( object ) :
        def __init__ ( self , stream ) :
            self.stream = stream
            self.stream = stream
        def has_next_byte ( self ) :
            if self.stream.read ( 1 ) < self.stream.read ( 1 ) :
                return True
            else :
                p = 0
                try :
                    length = stream.read ( self.stream.read ( self.stream.read ( self.stream.read ( self.stream.read ( self.stream.read ( self.stream.read ( self.stream.read ( self.stream.read ( self.stream.read ( self.stream.read ( self.stream.read ( self.stream.read ( self.stream.read ( self.stream.read ( self.stream.read ( self.stream.read ( stream.stream.read ( stream.stream.read ( stream.stream.stream.read ( stream.stream.stream.stream.read ( stream.stream.stream.stream.stream ) ) )